types/underscore/index.d.ts

```diff
@@ -254,58 +254,30 @@ declare module _ {
          * @param context `this` object in `iteratee`, optional.
          * @returns The reduced result.
          **/
-        reduce<V extends Collection<any>>(
+        reduce<V extends List<any>, TResult = TypeOfList<V>>(
             collection: V,
-            iteratee: MemoCollectionIterator<TypeOfCollection<V>, TypeOfCollection<V>, V>,
+            iteratee: MemoIterator<TypeOfList<V>, TResult | TypeOfList<V>, V>,
             memo?: undefined,
             context?: any
-        ): TypeOfCollection<V>;
-
-        /**
-         * Also known as inject and foldl, reduce boils down a collection of values into a
-         * single value. Memo is the initial state of the reduction, and each successive
-         * step of it should be returned by iteratee. The iteratee is passed four arguments:
-         * the memo, then the value and index (or key) of the iteration, and finally a reference
-         * to the entire collection.
-         *
-         * If no memo is passed to the initial invocation of reduce, the iteratee is not invoked
-         * on the first element of the collection. The first element is instead passed as the memo
-         * in the invocation of the iteratee on the next element in the collection.
-         * @param collection Reduces the elements of this collection.
-         * @param iteratee Reduce iteratee function for each element in `collection`.
-         * @param memo Initial reduce state or undefined to use the first collection item as initial state.
-         * @param context `this` object in `iteratee`, optional.
-         * @returns The reduced result.
-         **/
+        ): TResult | TypeOfList<V> | undefined;
```

Comment:
I must repeat that this signature does not exist in Underscore. Either, you pass fewer than three arguments and you get the "use the first element as the initial memo" behavior, or you pass anything including `undefined` as the memo and it will be used as such.

Reply:
I agree that this signature as I wrote it does not exist, but I still maintain that there are two separate signatures that should ideally be accounted for. I've removed `memo` and `context` altogether from this one and think that it is correct now; let me know if I missed anything though.

As you pointed out, I messed up by adding `undefined` to the case that takes a `memo`, but I think that makes this case even more interesting because it covers two special-case results instead of just one - undefined (for an empty collection) and the collection item type (for a collection with a single item in it).